### PR TITLE
Add starttls smtp authentication

### LINKED ISSUE #4272: Cannot configure smtp for outgoing email through Office365
Hi all,

I'm try to configure smtp with office 365 from a docker container. 

I'm using the currently latest version 0.22.2 on docker 1.13.0 running on ubuntu 14.04. If I set any smtp security I cannot save the settings and I obtain the error "Sorry, something went wrong. Please try again."
The mail server is reachable by the container. 

The container logs are:

01-31 07:50:33 ERROR metabase.email :: Error testing SMTP connection: 
01-31 07:50:33 WARN api.email :: Problem connecting to mail server: 
01-31 07:50:33 ERROR metabase.middleware :: PUT /api/email 500 (2 s) (0 DB calls)
{:message "Sorry, something went wrong.  Please try again."}

Is there a way to delete mail setting after save? I try without security which is wrong but now I cannot add new users because the system wants to invite them by mail (which do not work)

Thanks,
Nicola

### PR DESCRIPTION
**Pull request by thearthur**
Resolves #4272

Adds the starttls authentication option, and more importantly, attempts to detect when to use it.
If the initial connection to an outgoing mail server fails, then it will try each of the available security settings in order of security. The current preference order is: TLS, STARTTLS,SSL. It will not silently downgrade to a plaintext connection.

### CODE REVIEW

File: src/metabase/email.clj [hunk 1]
```diff
@@ -117,3 +109,37 @@
       (log/error "Error testing SMTP connection:" (.getMessage e))
       {:error   :ERROR
        :message (.getMessage e)})))
+
+(def ^:private email-security-order ["tls" "starttls" "ssl"])
+
+(defn- guess-smtp-security [details]
+  (loop [[security-type & more-to-try] email-security-order]
```
Comment (salsakran) [inline src/metabase/email.clj]: Can we add a comment to this function describing our approach here?

Comment (thearthur) [inline src/metabase/email.clj]: done

**Comment by thearthur** [comment]
Resolves #4272

**Comment by tlrobinson** [comment]
@thearthur I think the "resolves" need to go in the top description

**Review by salsakran — CHANGES_REQUESTED** [review]
Other than whitespace nitpicking, looks good to me! I don't have enough smtp accounts lying around to properly test it however.

**Pull request merged**